File: playwright-code-profiling/playwright-code-profiling-optimized.js
const { chromium } = require("playwright");

const MAX_CONCURRENCY = 5;

/** dummy function to simulate extract data
 @param {Page} page
*/
async function extractData(page) {
  console.log(`Extracting data from ${page.url()}`);
  // we are using sleep instead of extracting all the data
  await new Promise((r) => setTimeout(r, 2000));
}

/** Process a single product page
 * @param {BrowserContext} context
 * @param {String} url
 */
async function processProductPage(context, url) {
  let page = await context.newPage();
  try {
    await page.goto(url);
  } catch (r) {
    // retry once again to account for any network errors
    await page.goto(url);
  }
  await extractData(page);
  await page.close();
}

/** Processes a single listing page
 * @param {BrowserContext} context
 * @param {String} url
 */
async function processListingPage(context, url) {
  let products,
    productAnchorElems,
    productUrls = [];
  let page = await context.newPage();

  await page.goto(url, { timeout: 60000 });
  products = page.locator("li.product a[class*=product__link]");
  productAnchorElems = await products.elementHandles();
  for (index in productAnchorElems) {
    productUrls.push(await productAnchorElems[index].getAttribute("href"));
  }
  await page.close();
  return productUrls;
}

/** Process a batch of URLs simulataneously
 * @param {Function} task
 * @param {BrowserContext} context
 * @param {Array<String>} urls
 */
async function processBatches(task, context, urls) {
  let batches = [],
    results = [],
    promises;

  // splitting urls into batches
  for (let i = 0; i < urls.length; i += MAX_CONCURRENCY) {
    batches.push(urls.slice(i, i + MAX_CONCURRENCY));
  }
  for (let batchIdx in batches) {
    // resetting promises
    promises = [];
    // populating promises
    for (let argIdx in batches[batchIdx]) {
      let url = batches[batchIdx][argIdx];
      promises.push(task(context, url));
    }
    // waiting for batch to finish
    await Promise.all(promises).then((values) => results.push(...values));
  }
  return results;
}

async function intercept(route) {
  const unwantedResources = ["stylesheet", "image", "fonts", "script"];
  if (unwantedResources.includes(route.request().resourceType())) {
    await route.abort();
  } else {
    await route.continue();
  }
}

(async () => {
  const browser = await chromium.launch({
    headless: true,
  });
  const context = await browser.newContext();
  await context.route("**/*", intercept);

  let maxPagination = 5;

  // initializing the listingUrls with the start URL
  let listingUrls = ["https://scrapeme.live/shop/"];
  // populating the listing urls with the rest of pagination urls
  for (let pageNo = 2; pageNo < maxPagination + 2; pageNo++) {
    listingUrls.push(`https://scrapeme.live/shop/page/${pageNo}`);
  }

  // process listing page data
  let results = await processBatches(processListingPage, context, listingUrls);

  let productUrls = [];
  // results will be list of list of productUrls
  productUrls = productUrls.concat(...results);

  // process product pages
  await processBatches(processProductPage, context, productUrls);

  // ---------------------
  await context.close();
  await browser.close();
})();

File: playwright-code-profiling/playwright-code-profiling-optimized.py
import asyncio
import itertools
from playwright.async_api import Playwright, async_playwright


async def extract_data(page):
    print(f"Extracting data for {page.url}")
    # we are using sleep instead of extracting all the data
    await asyncio.sleep(2)


async def process_product_page(context, url, sem):
    """Extract data from a single product page"""
    async with sem:
        page = await context.new_page()
        try:
            await page.goto(url, timeout=60000)
        except Exception:
            # Retrying once to account for network errors
            await page.goto(url, timeout=60000)
        await extract_data(page)
        await page.close()


async def process_listing_page(context, url, sem):
    """Extract data from a single listing page"""
    async with sem:
        page = await context.new_page()
        await page.goto(url, timeout=60000)
        products = page.locator('li.product a[class*=product__link]')
        product_anchor_elements = await products.element_handles()
        product_urls = await asyncio.gather(*[
            i.get_attribute('href') for i in product_anchor_elements
        ])
        await page.close()
        return product_urls


async def intercept(route):
    if route.request.resource_type in {"stylesheet", 'image', 'fonts', 'script'}:
        await route.abort()
    else:
        await route.continue_()


async def run(playwright: Playwright) -> None:
    browser = await playwright.chromium.launch()
    context = await browser.new_context()
    
    # request interception
    await context.route('**/*', intercept)

    sem = asyncio.Semaphore(5)

    max_pagination = 5
    # list is initialised with starting url
    listing_urls = ["https://scrapeme.live/shop/"]
    for page_no in range(2, max_pagination + 2):
        listing_urls.append(f"https://scrapeme.live/shop/page/{page_no}")

    product_urls = []
    listing_coro = []
    for url in listing_urls:
        listing_coro.append(process_listing_page(context, url, sem))
    # product urls will be a list containing list of urls
    product_urls = await asyncio.gather(*listing_coro)

    product_coros = []
    # using itertools.chain to loop over product urls
    for url in itertools.chain(*product_urls):
        product_coros.append(process_product_page(context, url, sem))

    await asyncio.gather(*product_coros)

    # ---------------------
    await context.close()
    await browser.close()


async def main():
    async with async_playwright() as playwright:
        await run(playwright) 


asyncio.run(main())
